torchtune/models/clip/_position_embeddings.py

```diff
@@ -0,0 +1,167 @@
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+
+import torch
+from torch import nn
+
+# TODO (Felipe): add load hooks + interpolation on positional encodings,
+# so max_num_tiles can be variable and a trained model can be adapted to a
+# new value.
+
+
+class TokenPositionalEmbedding(nn.Module):
+    """
+    Token positional embedding for images. Different for every token (patch) in an image.
+
+    Args:
+        embed_dim (int): The dimensionality of each token embedding.
+        patch_grid_size (int): The side of a squared grid that represents how many
+             patches are in one tile, i.e. tile_size // patch_size.
+    """
+
+    def __init__(self, embed_dim: int, patch_grid_size: int) -> None:
+        super().__init__()
+        scale = embed_dim**-0.5
+        self.positional_embedding = nn.Parameter(
+            scale
+            * torch.randn((patch_grid_size**2 + 1, embed_dim))  # +1 for CLS token
+        )
+
+    def forward(self, x: torch.Tensor, *args) -> torch.Tensor:
+        """
+        args:
+            x (torch.Tensor): Tensor with shape (*, n_tokens, embed_dim)
+        returns:
+            torch.Tensor: The input tensor with added positional embeddings.
+        """
+        return x + self.positional_embedding
+
+
+class TiledTokenPositionalEmbedding(nn.Module):
+    """
+    Token positional embedding for tiled images. There are two positional embeddings in this module:
+        - local_token_positional_embedding: same for every tile, different for every token. Equivalent to
+            ``torchtune.models.clip._position_embeddings.TokenPositionalEmbedding``, but gated.
+        - global_token_positional_embedding: different for every tile, different for every token.
+
+    Notice that tile is different from patch (token). An image is divided into tiles during pre-processing,
+    and patches are the outcome of the convolution in the ViT applied to each tile.
+
+    Args:
+        max_num_tiles (int): The maximum number of tiles an image can be divided into.
+        embed_dim (int): The dimensionality of each token embedding.
+        patch_grid_size (int): The side of a squared grid that represents how many
+             patches are in one tile, i.e. tile_size // patch_size.
+    """
+
+    def __init__(
+        self, max_num_tiles: int, embed_dim: int, patch_grid_size: int
+    ) -> None:
+        super().__init__()
+        self.n_tokens_per_tile = patch_grid_size**2 + 1  # +1 for cls token
+        scale = embed_dim**-0.5
+
+        # different for every token, same for every tile
+        self.local_token_positional_embedding = nn.Parameter(
+            scale * torch.randn(patch_grid_size**2 + 1, embed_dim)
+        )
+
+        # different for every token, different for every tile
+        self.global_token_positional_embedding = nn.Parameter(
```

Comment:
ok looking at the code below I don't fully understand why this isn't a ```nn.Embedding```?

Reply:
I dont think that nn.Embedding would work for a 2d grid. Copying below my explanation from another post. Maybe i should add this to the docstring:

Imagine you have an image 400x800, tile_size = 200, aspect_ratio = (2,4).

num_tiles_h = 2
num_tiles_w = 4
image = torch.arange(num_tiles_h*num_tiles_w).reshape(num_tiles_h, num_tiles_w)
print(image)
print(image.reshape(-1))

# tensor([[0, 1, 2, 3],
#         [4, 5, 6, 7]]),
# tensor([0, 1, 2, 3, 4, 5, 6, 7]))

Each tile would have a different positional embedding. Our positional embedding supports max_num_tiles. Lets say max_num_tiles = 36.

pos_emb = torch.arange(36).reshape(6,6)
# tensor([[ 0,  1,  2,  3,  4,  5],
#         [ 6,  7,  8,  9, 10, 11],
#         [12, 13, 14, 15, 16, 17],
#         [18, 19, 20, 21, 22, 23],
#         [24, 25, 26, 27, 28, 29],
#         [30, 31, 32, 33, 34, 35]])

The code is supposed to add the positional embedding to the tile, so:
image = image + pos_emb[:num_tiles_h, :num_tiles_w]
tensor([0, 1, 2, 3, 4, 5, 6, 7])) + tensor([0, 1, 2, 3, 6, 7, 8, 9]))

Reply:
Thanks for the detailed clarification!